File: lib/embulk/command/embulk_run.rb
require 'embulk'

module Embulk
  def self.run(argv)
    # reset context class loader set by org.jruby.Main.main to nil. embulk manages
    # multiple classloaders. default classloader should be Plugin.class.getClassloader().
    java.lang.Thread.current_thread.set_context_class_loader(nil)

    require 'embulk/version'

    i = argv.find_index {|arg| arg !~ /^\-/ }
    unless i
      if argv.include?('--version')
        puts "embulk #{Embulk::VERSION}"
        system_exit_success
      end
      usage nil
    end
    subcmd = argv.slice!(i).to_sym

    require 'java'
    require 'optparse'
    op = OptionParser.new
    op.version = Embulk::VERSION

    puts "#{Time.now.strftime("%Y-%m-%d %H:%M:%S.%3N %z")}: Embulk v#{Embulk::VERSION}"

    plugin_paths = []
    load_paths = []
    classpaths = []
    classpath_separator = java.io.File.pathSeparator

    options = {
      system_config: {}
    }

    java_embed_ops = lambda do
      op.separator ""
      op.separator "  Other options:"
      op.on('-l', '--log PATH', 'Output log messages to a file (default: -)') do |path|
        options[:system_config][:log_path] = path
      end
      op.on('-l', '--log-level LEVEL', 'Log level (error, warn, info, debug or trace)') do |level|
        options[:system_config][:log_level] = level
      end
      op.on('-X KEY=VALUE', 'Add a performance system config') do |kv|
        k, v = kv.split('=', 2)
        v ||= "true"
        options[:system_config][k] = v
      end
    end

    plugin_load_ops = lambda do
      op.separator ""
      op.separator "  Plugin load options:"
      op.on('-L', '--load PATH', 'Add a local plugin path') do |plugin_path|
        plugin_paths << plugin_path
      end
      op.on('-I', '--load-path PATH', 'Add ruby script directory path ($LOAD_PATH)') do |load_path|
        load_paths << load_path
      end
      op.on('-C', '--classpath PATH', "Add java classpath separated by #{classpath_separator} (CLASSPATH)") do |classpath|
        classpaths.concat classpath.split(classpath_separator)
      end
      op.on('-b', '--bundle BUNDLE_DIR', 'Path to a Gemfile directory (create one using "embulk mkbundle" command)') do |path|
        # only for help message. implemented at lib/embulk/command/embulk_bundle.rb
      end
    end

    case subcmd
    when :run
      op.banner = "Usage: embulk run <config.yml>"
      op.separator "  Options:"
      op.on('-r', '--resume-state PATH', 'Path to a file to write or read resume state') do |path|
        options[:resume_state_path] = path
      end
      op.on('-o', '--output PATH', '(deprecated)') do |path|
        STDERR.puts "#{Time.now.strftime("%Y-%m-%d %H:%M:%S.%3N %z")}: Run with -o option is deprecated. Please use -c option instead. For example,"
        STDERR.puts "#{Time.now.strftime("%Y-%m-%d %H:%M:%S.%3N %z")}: "
        STDERR.puts "#{Time.now.strftime("%Y-%m-%d %H:%M:%S.%3N %z")}:   $ embulk run config.yml -c diff.yml"
        STDERR.puts "#{Time.now.strftime("%Y-%m-%d %H:%M:%S.%3N %z")}: "
        STDERR.puts "#{Time.now.strftime("%Y-%m-%d %H:%M:%S.%3N %z")}: This -c option stores only diff of the next configuration."
        STDERR.puts "#{Time.now.strftime("%Y-%m-%d %H:%M:%S.%3N %z")}: The diff will be merged to the original config.yml file."
        STDERR.puts "#{Time.now.strftime("%Y-%m-%d %H:%M:%S.%3N %z")}: "
        options[:next_config_output_path] = path
      end
      op.on('-c', '--config-diff PATH', 'Path to a file to read & write the next configuration diff') do |path|
        options[:next_config_diff_path] = path
      end
      plugin_load_ops.call
      java_embed_ops.call
      args = 1..1

    when :cleanup
      op.banner = "Usage: embulk cleanup <config.yml>"
      op.separator "  Options:"
      op.on('-r', '--resume-state PATH', 'Path to a file to cleanup resume state') do |path|
        options[:resume_state_path] = path
      end
      plugin_load_ops.call
      java_embed_ops.call
      args = 1..1

    when :preview
      op.banner = "Usage: embulk preview <config.yml>"
      op.separator "  Options:"
      op.on('-G', '--vertical', "Use vertical output format", TrueClass) do |b|
        options[:format] = "vertical"
      end
      plugin_load_ops.call
      java_embed_ops.call
      args = 1..1

    when :guess
      op.banner = "Usage: embulk guess <partial-config.yml>"
      op.separator "  Options:"
      op.on('-o', '--output PATH', 'Path to a file to write the guessed configuration') do |path|
        options[:next_config_output_path] = path
      end
      op.on('-g', '--guess NAMES', "Comma-separated list of guess plugin names") do |names|
        (options[:system_config][:guess_plugins] ||= []).concat names.split(",")  # TODO
      end
      plugin_load_ops.call
      java_embed_ops.call
      args = 1..1

    when :mkbundle
      op.banner = "Usage: embulk mkbundle <directory> [--path PATH]"
      op.separator "  Options:"
      op.on('--path PATH', 'Relative path from <directory> for the location to install gems to (e.g. --path shared/bundle).') do |path|
        options[:bundle_path] = path
      end
      op.separator <<-EOF

  "mkbundle" creates a new a plugin bundle directory. You can install
  plugins (gems) to the directory instead of ~/.embulk.

  See generated <directory>/Gemfile to install plugins to the directory.
  Use -b, --bundle BUNDLE_DIR option to use it:

    $ embulk mkbundle ./dir                # create bundle directory
    $ (cd dir && vi Gemfile && embulk bundle)   # update plugin list
    $ embulk guess -b ./dir ...            # guess using bundled plugins
    $ embulk run   -b ./dir ...            # run using bundled plugins

      EOF
      args = 1..1

    when :bundle
      if argv[0] == 'new'
        usage nil if argv.length != 2
        new_bundle(argv[1], nil)
        STDERR.puts "'embulk bundle new' is deprecated. This will be removed in future release. Please use 'embulk mkbundle' instead."
      else
        run_bundler(argv)
      end
      system_exit_success

    when :gem
      require 'rubygems/gem_runner'
      Gem::GemRunner.new.run argv
      system_exit_success

    when :new
      op.banner = "Usage: embulk new <category> <name>" + %[
categories:
    ruby-input                 Ruby record input plugin    (like "mysql")
    ruby-output                Ruby record output plugin   (like "mysql")
    ruby-filter                Ruby record filter plugin   (like "add-hostname")
    #ruby-file-input           Ruby file input plugin      (like "ftp")          # not implemented yet [#21]
    #ruby-file-output          Ruby file output plugin     (like "ftp")          # not implemented yet [#22]
    ruby-parser                Ruby file parser plugin     (like "csv")
    ruby-formatter             Ruby file formatter plugin  (like "csv")
    #ruby-decoder              Ruby file decoder plugin    (like "gzip")         # not implemented yet [#31]
    #ruby-encoder              Ruby file encoder plugin    (like "gzip")         # not implemented yet [#32]
    java-input                 Java record input plugin    (like "mysql")
    java-output                Java record output plugin   (like "mysql")
    java-filter                Java record filter plugin   (like "add-hostname")
    java-file-input            Java file input plugin      (like "ftp")
    java-file-output           Java file output plugin     (like "ftp")
    java-parser                Java file parser plugin     (like "csv")
    java-formatter             Java file formatter plugin  (like "csv")
    java-decoder               Java file decoder plugin    (like "gzip")
    java-encoder               Java file encoder plugin    (like "gzip")

examples:
    new ruby-output hbase
    new ruby-filter int-to-string
]
      args = 2..2

    when :migrate
      op.banner = "Usage: embulk migrate <directory>"
      args = 1..1

    when :selfupdate
      op.on('-f', "Skip corruption check", TrueClass) do |b|
        options[:force] = true
      end
      args = 0..1

    when :example
      args = 0..1

    when :exec
      exec(*argv)
      exit 127

    when :irb
      require 'irb'
      IRB.start
      system_exit_success

    else
      usage "Unknown subcommand #{subcmd.to_s.dump}."
    end

    begin
      op.parse!(argv)
      unless args.include?(argv.length)
        usage_op op, nil
      end
    rescue => e
      usage_op op, e.to_s
    end

    case subcmd
    when :example
      require 'embulk/command/embulk_example'
      path = ARGV[0] || "embulk-example"
      puts "Creating #{path} directory..."
      Embulk.create_example(path)
      puts ""
      puts "Run following subcommands to try embulk:"
      puts ""
      puts "   1. embulk guess #{File.join(path, 'seed.yml')} -o config.yml"
      puts "   2. embulk preview config.yml"
      puts "   3. embulk run config.yml"
      puts ""

    when :new
      lang_cate = ARGV[0]
      name = ARGV[1]

      language, category = case lang_cate
        when "java-input"       then [:java, :input]
        when "java-output"      then [:java, :output]
        when "java-filter"      then [:java, :filter]
        when "java-file-input"  then [:java, :file_input]
        when "java-file-output" then [:java, :file_output]
        when "java-parser"      then [:java, :parser]
        when "java-formatter"   then [:java, :formatter]
        when "java-decoder"     then [:java, :decoder]
        when "java-encoder"     then [:java, :encoder]
        when "ruby-input"       then [:ruby, :input]
        when "ruby-output"      then [:ruby, :output]
        when "ruby-filter"      then [:ruby, :filter]
        when "ruby-file-input"  then raise "ruby-file-input is not implemented yet. See #21 on github." #[:ruby, :file_input]
        when "ruby-file-output" then raise "ruby-file-output is not implemented yet. See #22 on github." #[:ruby, :file_output]
        when "ruby-parser"      then [:ruby, :parser]
        when "ruby-formatter"   then [:ruby, :formatter]
        when "ruby-decoder"     then raise "ruby-decoder is not implemented yet. See #31 on github." #[:ruby, :decoder]
        when "ruby-encoder"     then raise "ruby-decoder is not implemented yet. See #32 on github." #[:ruby, :encoder]
        else
          usage_op op, "Unknown category '#{lang_cate}'"
        end

      require 'embulk/command/embulk_new_plugin'
      Embulk.new_plugin(name, language, category)

    when :migrate
      path = ARGV[0]
      require 'embulk/command/embulk_migrate_plugin'
      Embulk.migrate_plugin(path)

    when :selfupdate
      require 'embulk/command/embulk_selfupdate'
      options[:version] = ARGV[0]
      Embulk.selfupdate(options)

    when :mkbundle
      new_bundle(argv[0], options[:bundle_path])

    else
      require 'json'

      # Gem::StubSpecification is an internal API that seems chainging often.
      # Gem::Specification.add_spec is deprecated also. Therefore, here makes
      # -L <path> option alias of -I <path>/lib by assuming that *.gemspec file
      # always has require_paths = ["lib"].
      load_paths = load_paths + plugin_paths.map {|path| File.join(path, "lib") }

      setup_load_paths(load_paths)
      setup_classpaths(classpaths)

      # call setup after setup_classpaths to allow users to overwrite
      # embulk classes
      Embulk.setup(options.delete(:system_config))

      begin
        case subcmd
        when :guess
          Embulk::Runner.guess(argv[0], options)
        when :preview
          Embulk::Runner.preview(argv[0], options)
        when :run
          Embulk::Runner.run(argv[0], options)
        end
      rescue => ex
        print_exception(ex)
        puts ""
        puts "Error: #{ex}"
        raise SystemExit.new(1, ex.to_s)
      end
    end
  end

  def self.default_gem_home
    if RUBY_PLATFORM =~ /java/i
      user_home = java.lang.System.properties["user.home"]
    end
    user_home ||= ENV['HOME']
    unless user_home
      raise "HOME environment variable is not set."
    end
    File.expand_path File.join(user_home, '.embulk', Gem.ruby_engine, RbConfig::CONFIG['ruby_version'])
  end

  private

  def self.setup_load_paths(load_paths)
    # first $LOAD_PATH has highet priority. later load_paths should have highest priority.
    load_paths.each do |load_path|
      # ruby script directory (use unshift to make it highest priority)
      $LOAD_PATH.unshift File.expand_path(load_path)
    end
  end

  def self.setup_classpaths(classpaths)
    classpaths.each do |classpath|
      $CLASSPATH << classpath  # $CLASSPATH object doesn't have concat method
    end
  end

  def self.new_bundle(path, bundle_path)
    require 'fileutils'
    require 'rubygems/gem_runner'

    if File.exists?(path)
      error = "'#{path}' already exists."
      STDERR.puts error
      raise SystemExit.new(1, error)
    end

    puts "Initializing #{path}..."
    FileUtils.mkdir_p path
    begin
      success = false

      # copy embulk/data/bundle/ contents
      require 'embulk/data/package_data'
      pkg = PackageData.new("bundle", path)
      %w[Gemfile .ruby-version .bundle/config embulk/input/example.rb embulk/output/example.rb embulk/filter/example.rb].each do |file|
        pkg.cp(file, file)
      end

      # run the first bundle-install
      Dir.chdir(path) do
        run_bundler(['install', '--path', bundle_path || "."])
      end

      success = true
    rescue Gem::SystemExitException => e
      raise e if e.exit_code != 0
      success = true
    ensure
      FileUtils.rm_rf path unless success
    end
  end

  def self.run_bundler(argv)
    require 'bundler'  # bundler is included in embulk-core.jar

    # this hack is necessary to make --help working
    Bundler.define_singleton_method(:which_orig, Bundler.method(:which))
    Bundler.define_singleton_method(:which) do |executable|
      if executable == "man"
        false
      else
        which_orig(executable)
      end
    end

    require 'bundler/friendly_errors'
    require 'bundler/cli'
    Bundler.with_friendly_errors do
      Bundler::CLI.start(argv, debug: true)
    end
  end

  def self.usage(message)
    STDERR.puts "Embulk v#{Embulk::VERSION}"
    STDERR.puts "Usage: embulk [-vm-options] <command> [--options]"
    STDERR.puts "Commands:"
    STDERR.puts "   mkbundle   <directory>                             # create a new plugin bundle environment."
    STDERR.puts "   bundle     [directory]                             # update a plugin bundle environment."
    STDERR.puts "   run        <config.yml>                            # run a bulk load transaction."
    STDERR.puts "   cleanup    <config.yml>                            # cleanup resume state."
    STDERR.puts "   preview    <config.yml>                            # dry-run the bulk load without output and show preview."
    STDERR.puts "   guess      <partial-config.yml> -o <output.yml>    # guess missing parameters to create a complete configuration file."
    STDERR.puts "   gem        <install | list | help>                 # install a plugin or show installed plugins."
    STDERR.puts "                                                      # plugin path is #{ENV['GEM_HOME']}"
    STDERR.puts "   new        <category> <name>                       # generates new plugin template"
    STDERR.puts "   migrate    <path>                                  # modify plugin code to use the latest Embulk plugin API"
    STDERR.puts "   example    [path]                                  # creates an example config file and csv file to try embulk."
    STDERR.puts "   selfupdate [version]                               # upgrades embulk to the latest released version or to the specified version."
    STDERR.puts ""
    STDERR.puts "VM options:"
    STDERR.puts "   -J-O                             Disable JVM optimizations to speed up startup time (enabled by default if command is 'run')"
    STDERR.puts "   -J+O                             Enable JVM optimizations to speed up throughput"
    STDERR.puts "   -J...                            Set JVM options (use -J-help to see available options)"
    STDERR.puts "   -R...                            Set JRuby options (use -R--help to see available options)"
    STDERR.puts ""
    if message
      system_exit "error: #{message}"
    else
      system_exit "Use \`<command> --help\` to see description of the commands."
    end
  end

  def self.usage_op(op, message)
    STDERR.puts op.help
    STDERR.puts
    system_exit message
  end

  def self.print_exception(ex)
    if ex.respond_to?(:to_java) && ex.is_a?(java.lang.Throwable)
      ex.to_java.printStackTrace(java.lang.System.out)
    else
      puts "#{ex.to_s}"
      ex.backtrace.each do |bt|
        puts "    #{bt}"
      end
    end
  end

  def self.system_exit(message=nil)
    STDERR.puts message if message
    raise SystemExit.new(1, message)
  end

  def self.system_exit_success
    raise SystemExit.new(0)
  end
end

File: embulk-core/src/main/java/org/embulk/cli/EmbulkRun.java
package org.embulk.cli;

import java.io.IOException;
import java.nio.file.FileVisitResult;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.nio.file.SimpleFileVisitor;
import java.nio.file.attribute.BasicFileAttributes;
import java.util.Arrays;
import java.util.List;
import java.util.Properties;
import org.embulk.EmbulkEmbed;
import org.embulk.EmbulkRunner;
import org.embulk.EmbulkSystemProperties;
import org.embulk.EmbulkVersion;
import org.embulk.jruby.LazyScriptingContainerDelegate;
import org.embulk.jruby.ScriptingContainerDelegate;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

public class EmbulkRun {
    public static int run(final CommandLine commandLine, final EmbulkSystemProperties embulkSystemProperties) {
        return (new EmbulkRun()).runInternal(commandLine, embulkSystemProperties);
    }

    private int runInternal(final CommandLine commandLine, final EmbulkSystemProperties embulkSystemProperties) {
        final List<String> subcommandArguments = commandLine.getArguments();

        switch (commandLine.getCommand()) {
            case EXAMPLE:
                final EmbulkExample embulkExample = new EmbulkExample();
                try {
                    embulkExample.createExample(commandLine.getArguments().isEmpty()
                                                        ? "embulk-example"
                                                        : commandLine.getArguments().get(0));
                } catch (IOException ex) {
                    ex.printStackTrace(System.err);
                    return 1;
                }
                return 0;
            case NEW:
                final String categoryWithLanguage = commandLine.getArguments().get(0);
                final String nameGiven = commandLine.getArguments().get(1);
                try {
                    final EmbulkNew embulkNew = new EmbulkNew(categoryWithLanguage, nameGiven, EmbulkVersion.VERSION);
                    embulkNew.newPlugin();
                } catch (IOException ex) {
                    ex.printStackTrace();
                    return 1;
                }
                return 0;
            case SELFUPDATE:
                if (commandLine.getArguments().isEmpty()) {
                    System.err.println("'embulk selfupdate' requires the target version since v0.10.0. It no longer updates to the latest version.");
                    return 1;
                }
                try {
                    SelfUpdate.toSpecific(
                            EmbulkVersion.VERSION,
                            commandLine.getArguments().get(0),
                            "true".equals(embulkSystemProperties.getProperty("force_selfupdate")));
                } catch (final IOException ex) {
                    ex.printStackTrace();
                    return 1;
                }
                return 0;
            case BUNDLE:
                if (!subcommandArguments.isEmpty() && subcommandArguments.get(0).equals("new")) {
                    System.err.println("embulk: 'embulk bundle new' is no longer available. Please use 'embulk mkbundle' instead.");
                    return -1;
                } else {
                    return runBundler(subcommandArguments, null, embulkSystemProperties);
                }
            case GEM:
                return callJRubyGem(subcommandArguments, embulkSystemProperties);
            case MKBUNDLE:
                newBundle(commandLine.getArguments().get(0), embulkSystemProperties);
                break;
            case RUN:
            case CLEANUP:
            case PREVIEW:
            case GUESS:
                // NOTE: When it was in Ruby "require 'embulk'" was required on top for Ruby
                // |Embulk::setup|.
                // Ruby |Embulk::setup| is now replaced with direct calls to EmbulkEmbed below.

                // TODO: Move this to initial JRuby instantiation.
                // reset context class loader set by org.jruby.Main.main to nil. embulk manages
                // multiple classloaders. default classloader should be
                // Plugin.class.getClassloader().
                Thread.currentThread().setContextClassLoader(null);

                // NOTE: When it was in Ruby ""require 'json'" was required.

                // NOTE: $LOAD_PATH and $CLASSPATH are set in JRubyInitializer via system config.

                // Ruby |Embulk::Runner| contained the EmbulkRunner instance, but it's no longer available.
                final EmbulkEmbed.Bootstrap bootstrap = new EmbulkEmbed.Bootstrap();
                bootstrap.setEmbulkSystemProperties(embulkSystemProperties);

                // see embulk-core/src/main/java/org/embulk/jruby/JRubyScriptingModule.
                final EmbulkRunner runner = new EmbulkRunner(bootstrap.initialize(), embulkSystemProperties);

                final Path configDiffPath = getPathFromProperties("config_diff_path", embulkSystemProperties);
                final Path outputPath = getPathFromProperties("output_path", embulkSystemProperties);
                final Path resumeStatePath = getPathFromProperties("resume_state_path", embulkSystemProperties);
                final String previewFormat = embulkSystemProperties.getProperty("preview_format");

                try {
                    switch (commandLine.getCommand()) {
                        case GUESS:
                            runner.guess(Paths.get(commandLine.getArguments().get(0)), outputPath);
                            break;
                        case PREVIEW:
                            runner.preview(Paths.get(commandLine.getArguments().get(0)), previewFormat);
                            break;
                        case RUN:
                            runner.run(Paths.get(commandLine.getArguments().get(0)),
                                       configDiffPath,
                                       outputPath,
                                       resumeStatePath);
                            break;
                        default:  // Do nothing, and just pass through.
                    }
                } catch (Throwable ex) {
                    ex.printStackTrace(System.err);
                    System.err.println("");
                    System.err.println("Error: " + ex.getMessage());
                    return 1;
                }
                break;
            default:  // Do nothing, and just pass through.
        }
        return 0;
    }

    private static Path getPathFromProperties(final String key, final Properties properties) {
        final String value = properties.getProperty(key);
        if (value == null) {
            return null;
        }
        return Paths.get(value);
    }

    private void copyResourceToFile(final String sourceResourceName,
                                    final Path destinationBasePath,
                                    final String destinationRelativePath) throws IOException {
        final Path destinationPath = destinationBasePath.resolve(destinationRelativePath);
        System.out.println("  Creating " + destinationRelativePath);
        Files.createDirectories(destinationPath.getParent());
        Files.copy(EmbulkRun.class.getClassLoader().getResourceAsStream(sourceResourceName), destinationPath);
    }

    private int newBundle(final String pathString, final EmbulkSystemProperties embulkSystemProperties) {
        final Path path = Paths.get(pathString).toAbsolutePath();
        if (Files.exists(path)) {
            System.err.println("'" + pathString + "' already exists.");
            return 1;
        }

        System.out.println("Initializing " + pathString + "...");
        try {
            Files.createDirectories(path);
        } catch (IOException ex) {
            ex.printStackTrace();
            return 1;
        }

        boolean success = false;
        try {
            // copy embulk/data/bundle/ contents
            copyResourceToFile("org/embulk/jruby/bundler/template/Gemfile", path, "Gemfile");
            // ".ruby-version" is no longer required since JRuby used is embedded in Embulk.
            copyResourceToFile("org/embulk/jruby/bundler/template/.bundle/config", path, ".bundle/config");
            copyResourceToFile("org/embulk/jruby/bundler/template/embulk/input/example.rb", path, "embulk/input/example.rb");
            copyResourceToFile("org/embulk/jruby/bundler/template/embulk/output/example.rb", path, "embulk/output/example.rb");
            copyResourceToFile("org/embulk/jruby/bundler/template/embulk/filter/example.rb", path, "embulk/filter/example.rb");
        } catch (IOException ex) {
            ex.printStackTrace();
            throw new RuntimeException(ex);
        }

        int result = -1;
        final String bundlePath = embulkSystemProperties.getProperty("bundle_path");
        try {
            // run the first bundle-install
            result = runBundler(
                    Arrays.asList("install", "--path", bundlePath != null ? bundlePath : "."), path, embulkSystemProperties);
            if (result == 0) {
                success = true;
            }
        } catch (Exception ex) {
            ex.printStackTrace();
            throw ex;
        } finally {
            if (!success) {
                try {
                    Files.walkFileTree(path, new SimpleFileVisitor<Path>() {
                            @Override
                            public FileVisitResult visitFile(Path file, BasicFileAttributes attributes) {
                                try {
                                    Files.deleteIfExists(file);
                                } catch (IOException ex) {
                                    // Ignore.
                                }
                                return FileVisitResult.CONTINUE;
                            }

                            @Override
                            public FileVisitResult postVisitDirectory(Path dir, IOException exception) {
                                try {
                                    Files.deleteIfExists(dir);
                                } catch (IOException ex) {
                                    // Ignore.
                                }
                                return FileVisitResult.CONTINUE;
                            }
                        });
                } catch (IOException ex) {
                    ex.printStackTrace();
                    return 1;
                }
            }
        }
        return result;
    }

    private int runBundler(final List<String> arguments, final Path path, final EmbulkSystemProperties embulkSystemProperties) {
        final ScriptingContainerDelegate localJRubyContainer;
        try {
            localJRubyContainer = createJRubyForRubyCommand(embulkSystemProperties, "bundle");
        } catch (final NullPointerException ex) {
            // TODO: Handle the exception better and have a better error message.
            System.err.println(ex.getMessage());
            return -1;
        }

        localJRubyContainer.runScriptlet("require 'bundler'");  // bundler is included in embulk-core.jar

        // this hack is necessary to make --help working
        localJRubyContainer.runScriptlet("Bundler.define_singleton_method(:which_orig, Bundler.method(:which))");
        localJRubyContainer.runScriptlet("Bundler.define_singleton_method(:which) { |executable| (executable == 'man' ? false : which_orig(executable)) }");

        localJRubyContainer.runScriptlet("require 'bundler/friendly_errors'");
        localJRubyContainer.runScriptlet("require 'bundler/cli'");

        localJRubyContainer.put("__internal_argv_java__", arguments);
        if (path == null) {
            localJRubyContainer.runScriptlet("Bundler.with_friendly_errors { Bundler::CLI.start(Array.new(__internal_argv_java__), debug: true) }");
        } else {
            localJRubyContainer.put("__internal_working_dir__", path.toString());
            localJRubyContainer.runScriptlet(
                    "Dir.chdir(__internal_working_dir__) {"
                    + "  Bundler.with_friendly_errors { Bundler::CLI.start(Array.new(__internal_argv_java__), debug: true) }"
                    + "}");
            localJRubyContainer.remove("__internal_working_dir__");
        }
        localJRubyContainer.remove("__internal_argv_java__");

        return 0;
    }

    // TODO: Check if it is required to process JRuby options.
    private static ScriptingContainerDelegate createJRubyForRubyCommand(
            final EmbulkSystemProperties embulkSystemProperties,
            final String command) {
        final String propertyGemHome = embulkSystemProperties.getProperty("gem_home");
        if (propertyGemHome == null || propertyGemHome.isEmpty()) {
            logger.error("Embulk system property \"gem_home\" is not configured.");
            throw new NullPointerException("Embulk system property \"gem_home\" is not configured.");
        }
        final String propertyGemPath = embulkSystemProperties.getProperty("gem_path");  // gem_path is optional.

        final ScriptingContainerDelegate jruby = LazyScriptingContainerDelegate.withGemsIgnored(logger, embulkSystemProperties);

        if (jruby == null) {
            // TODO: Handle the exception better and have a better error message.
            throw new NullPointerException(
                    "JRuby is not configured well to run \"" + command + "\". Configure the Embulk system property \"jruby\".");
        }

        // The environment variables "GEM_HOME" (and "GEM_PATH") are mandatory for the "gem" and "bundle" commands.
        //
        // Unlike normal RubyGem loading, `Gem.paths.home=` does not work for the "gem" and "bundle" commands
        //
        // `Gem.paths` is overwritten in `GemRunner.new` in "rubygems".
        // https://github.com/rubygems/rubygems/blob/v3.1.4/lib/rubygems/gem_runner.rb#L80

        final String currentGemHome;
        if (jruby.callMethod(jruby.runScriptlet("ENV"), "has_key?", "GEM_HOME", Boolean.class)) {
            currentGemHome = jruby.callMethod(jruby.runScriptlet("ENV"), "fetch", "GEM_HOME", String.class);
        } else {
            currentGemHome = null;
        }
        if (currentGemHome == null || currentGemHome.isEmpty()) {
            logger.info(
                    "Environment variable \"GEM_HOME\" is not set. "
                    + "Setting \"GEM_HOME\" to \"{}\" from Embulk system property \"gem_home\" for the \"{}\" command.",
                    propertyGemHome, command);
            jruby.callMethod(jruby.runScriptlet("ENV"), "store", "GEM_HOME", propertyGemHome);
        } else if (!currentGemHome.equals(propertyGemHome)) {
            logger.info(
                    "Environment variable \"GEM_HOME\" is \"{}\", "
                    + "which is different from Embulk system property \"gem_home\". "
                    + "Resetting \"GEM_HOME\" to \"{}\" from \"gem_home\" for the \"{}\" command.",
                    currentGemHome, propertyGemHome, command);
            jruby.callMethod(jruby.runScriptlet("ENV"), "store", "GEM_HOME", propertyGemHome);
        }

        final String currentGemPath;
        if (jruby.callMethod(jruby.runScriptlet("ENV"), "has_key?", "GEM_PATH", Boolean.class)) {
            currentGemPath = jruby.callMethod(jruby.runScriptlet("ENV"), "fetch", "GEM_PATH", String.class);
        } else {
            currentGemPath = null;
        }
        jruby.callMethod(jruby.runScriptlet("ENV"), "store", "GEM_PATH", embulkSystemProperties.getProperty("gem_path"));
        if (currentGemPath == null || currentGemPath.isEmpty()) {
            if (propertyGemPath != null && !propertyGemPath.isEmpty()) {
                logger.info(
                        "Environment variable \"GEM_PATH\" is not set while Embulk system property \"gem_path\" is set. "
                        + "Setting \"GEM_PATH\" to \"{}\" from \"gem_path\" for the \"{}\" command.",
                        propertyGemPath, command);
                jruby.callMethod(jruby.runScriptlet("ENV"), "store", "GEM_PATH", propertyGemPath);
            }
        } else if (!currentGemPath.equals(propertyGemPath)) {
            if (propertyGemPath == null && propertyGemPath.isEmpty()) {
                logger.info(
                        "Environment variable \"GEM_PATH\" is \"{}\" while Embulk system property \"gem_path\" is not set. "
                        + "Unsetting \"GEM_PATH\" from \"gem_path\" for the \"{}\" command.",
                        currentGemPath, command);
                jruby.callMethod(jruby.runScriptlet("ENV"), "delete", "GEM_PATH");
            } else {
                logger.info(
                        "Environment variable \"GEM_PATH\" is \"{}\", "
                        + "which is different from Embulk system property \"gem_path\". "
                        + "Resetting \"GEM_PATH\" to \"{}\" from \"gem_path\" for the \"{}\" command.",
                        currentGemPath, propertyGemPath, command);
                jruby.callMethod(jruby.runScriptlet("ENV"), "store", "GEM_PATH", propertyGemPath);
            }
        }

        // NOTE: Same done in JRubyScriptingModule.
        // Remember to update |org.embulk.jruby.JRubyScriptingModule| when these environment variables are changed.
        if (jruby.callMethod(jruby.runScriptlet("ENV"), "has_key?", "BUNDLE_GEMFILE", Boolean.class)) {
            final String currentBundleGemFile =
                    jruby.callMethod(jruby.runScriptlet("ENV"), "fetch", "BUNDLE_GEMFILE", String.class);
            logger.warn(
                    "Environment variable \"BUNDLE_GEMFILE\" is set. Unsetting \"BUNDLE_GEMFILE\" for the \"{}\" command.",
                    command);
            jruby.callMethod(jruby.runScriptlet("ENV"), "delete", "BUNDLE_GEMFILE");
        }

        return jruby;
    }

    private int callJRubyGem(final List<String> subcommandArguments, final EmbulkSystemProperties embulkSystemProperties) {
        final ScriptingContainerDelegate localJRubyContainer;
        try {
            localJRubyContainer = createJRubyForRubyCommand(embulkSystemProperties, "gem");
        } catch (final NullPointerException ex) {
            // TODO: Handle the exception better and have a better error message.
            System.err.println(ex.getMessage());
            return -1;
        }

        localJRubyContainer.runScriptlet("require 'rubygems/gem_runner'");
        localJRubyContainer.put("__internal_argv_java__", subcommandArguments);
        localJRubyContainer.runScriptlet("Gem::GemRunner.new.run Array.new(__internal_argv_java__)");
        localJRubyContainer.remove("__internal_argv_java__");

        return 0;
    }

    private static final Logger logger = LoggerFactory.getLogger(EmbulkRun.class);
}
